tests/FunctionalJavascript/SettingsFormTest.php

```diff
@@ -0,0 +1,273 @@
+use Drupal\FunctionalJavascriptTests\WebDriverTestBase;
+use Drupal\oe_authentication\Form\AuthenticationSettingsForm;
+
+/**
+ * Tests the settings form.
+ */
+class SettingsFormTest extends WebDriverTestBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected $defaultTheme = 'stark';
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'oe_authentication_test',
+  ];
+
+  /**
+   * Tests the two-factor authentication condition settings.
+   */
+  public function testTwoFactorAuthenticationConditionSettings(): void {
+    $role_id = $this->createRole(['administer authentication configuration']);
+    $this->drupalLogin($this->createUser(values: ['roles' => [$role_id]]));
+    $this->drupalGet('/admin/config/system/oe_authentication');
+
+    $assert_session = $this->assertSession();
+    $use_2fa_fieldset = $assert_session->elementExists('named', ['fieldset', 'Use two-factor authentication']);
+    $this->assertCount(3, $use_2fa_fieldset->findAll('css', 'input'));
+    $never_radio = $assert_session->fieldExists('Never', $use_2fa_fieldset);
+    $this->assertTrue($never_radio->isChecked());
+    $always_radio = $assert_session->fieldExists('Always', $use_2fa_fieldset);
+    $this->assertFalse($always_radio->isChecked());
+    $conditional_radio = $assert_session->fieldExists('Conditional', $use_2fa_fieldset);
+    $this->assertFalse($conditional_radio->isChecked());
+
+    $conditions_wrapper = $this->assertDetailsElementExists('Two-factor authentication conditions');
+    $this->assertFalse($conditions_wrapper->isVisible());
+    $assert_session->buttonExists('Save configuration')->press();
+    $assert_session->statusMessageContains('The configuration options have been saved.', 'status');
+
+    // The "never" option sets the "force_2fa" to FALSE.
+    $config = $this->loadConfig();
+    $this->assertEquals([], $config->get('2fa_conditions'));
+    $this->assertFalse($config->get('force_2fa'));
+
+    // Re-assert that the conditions are still hidden.
+    $this->drupalGet('/admin/config/system/oe_authentication');
+    $this->assertFalse($conditions_wrapper->isVisible());
+    // Set the 2FA to "always".
+    $always_radio->click();
+    $this->assertFalse($conditions_wrapper->isVisible());
+    $assert_session->buttonExists('Save configuration')->press();
+    $assert_session->statusMessageContains('The configuration options have been saved.', 'status');
+
+    // The "always" option sets the "force_2fa" to TRUE.
+    $config = $this->loadConfig();
+    $this->assertEquals([], $config->get('2fa_conditions'));
+    $this->assertTrue($config->get('force_2fa'));
+
+    $this->drupalGet('/admin/config/system/oe_authentication');
+    $this->assertFalse($conditions_wrapper->isVisible());
+    $conditional_radio->click();
+    $conditions_wrapper_selector = 'details[data-drupal-selector="edit-2fa-conditions"]';
+    $conditions_wrapper = $assert_session->waitForElementVisible('css', $conditions_wrapper_selector);
```

Comment:
What's a bit interesting: Here we wait until it is visible.
But above when we assert it is not visible after a click, we don't wait.
So in theory it could become visible above 10 milliseconds after clicking "Always".

(There is nothing we can do here, just an observation. You can resolve.)

Reply:
To be honest the wait it's not really needed as there's no animation, and JS tests have implicit waits.
But the main reason for not waiting for the method to be hidden is that I don't have an helper ready.

Reply:
the main problem is we would always have to wait the maximum time to verify that something does _not_ happen. So yeah nothing we can do, it just means the asserts above may not be full reliable.